samcli/lib/pipeline/bootstrap/resource.py

```diff
@@ -0,0 +1,131 @@
+
+    Attributes
+    ----------
+    partition: str
+        the partition part(AWS, aws-cn or aws-us-gov)  of the ARN
+    service: str
+        the service part(S3, IAM, ECR, ...etc) of the ARN
+    region: str
+        the AWS region part(us-east-1, eu-west-1, ...etc) of the ARN
+    account-id: str
+        the account-id part of the ARN
+    resource-id: str
+        the resource-id part of the ARN
+    resource-type: str
+        the resource-type part of the ARN
+    """
+
+    partition: str
+    service: str
+    region: str
+    account_id: str
+    resource_id: str
+
+    def __init__(self, arn: str) -> None:
+        parts = arn.split(":")
+        try:
+            [_, self.partition, self.service, self.region, self.account_id, self.resource_id] = parts
+        except ValueError as ex:
+            raise ValueError(f"Invalid ARN ({arn})") from ex
+
+
+class Resource:
+    """
+    Represents an AWS resource
+
+    Attributes
+    ----------
+    arn: str
+        the ARN of the resource
+    is_user_provided: bool
+        True if the user provided the ARN of the resource during the initialization. It indicates whether this pipeline-
+        resource is provided by the user or created by SAM during `sam pipeline bootstrap`
+
+    Methods
+    -------
+    name(self) -> Optional[str]:
+        extracts and returns the resource name from its ARN
+    """
+
+    def __init__(self, arn: Optional[str]) -> None:
+        self.arn: Optional[str] = arn
+        self.is_user_provided: bool = bool(arn)
+
+    def name(self) -> Optional[str]:
+        """
+        extracts and returns the resource name from its ARN
+        Raises
+        ------
+        ValueError if the ARN is invalid
+        """
+        if not self.arn:
+            return None
+        arn_parts: ARNParts = ARNParts(arn=self.arn)
+        return arn_parts.resource_id
+
+
+class IamUser(Resource):
+    """
+    Represents an AWS IamUser resource
+    Attributes
+    ----------
+    access_key_id: Optional[str]
+        holds the AccessKeyId of the credential of this IAM user, if any.
+    secret_access_key: Optional[str]
+        holds the SecretAccessKey of the credential of this IAM user, if any.
+    """
+
+    def __init__(
+        self, arn: Optional[str], access_key_id: Optional[str] = None, secret_access_key: Optional[str] = None
+    ) -> None:
+        self.access_key_id: Optional[str] = access_key_id
+        self.secret_access_key: Optional[str] = secret_access_key
+        super().__init__(arn=arn)
+
+
+class S3Bucket(Resource):
+    """
+    Represents an AWS S3Bucket resource
+    Attributes
+    ----------
+    kms_key_arn: Optional[str]
+        The ARN of the KMS key used in encrypting this S3Bucket, if any.
+    """
+
+    def __init__(self, arn: Optional[str], kms_key_arn: Optional[str] = None) -> None:
+        self.kms_key_arn: Optional[str] = kms_key_arn
+        super().__init__(arn=arn)
+
+
+class EcrRepo(Resource):
```

Comment:
Nit: ECRRepo instead of EcrRepo. Same goes for other class names.
> Note: When using abbreviations in CapWords, capitalize all the letters of the abbreviation. Thus HTTPServerError is better than HttpServerError.

Reply:
I agree on that but we are already use the CamelCase with the abbreviations in many other places like for example:
AWS::Serverless::Api and AWS::Serverless::HttpApi

Reply:
This is for Python naming convention instead of CFN. https://legacy.python.org/dev/peps/pep-0008/#descriptive-naming-styles

Reply:
I see your point, fixed.